### PR TITLE
Feature/auth

### PR DESCRIPTION
**Pull request by nein-tanuki**
### Models

- Added Admin
    - email: string
    - password : string (encrypted)

- Added User
    - email: string
    - password : string (encrypted)

- Added JwtDenyList
    Stores revoked jwt tokens
     - jti: string
     - exp: datetime

- Enabled uuid generation on primary keys of models Admin and User

### Controllers

- Added `registrations` and `sessions` controller for both user and admin
- Added `members` controller

#### Notes:
 - The members controller is generated to test protected endpoints. It will be removed later
 - **[IMPORTANT]** In registrations and sessions controller (both user and admin), you can find in **line 3** the code `skip_before_action :verify_authenticity_token`. This is used to prevent rails from verifying csrf tokens that's served on the front-end of our application. Since, the application has no front-end yet, it is advisable to disable this feature. It will be enabled once the front-end is done.

### Routes

- Added the following routes:

    `users` - registers user
    `users/sign_in` - signs in user
    `users/sign_out` - signs out user
    `admin` - registers admin
    `admin/sign_in` - signs in admin
    `admin/sign_out` - signs out admin
    `members-data` - get members endpoint

### Updated Gems/Packages

- Added the following gems/packages
```
gem 'devise-jwt'
gem 'dotenv-rails', groups: [:development, :test]
gem 'rack-cors'
```

### Development

- Added .env.example file

### Bugs/Issues

None

### Screenshots
localhost_5100_user_register - Google Chrome 07_03_2022 6_59_56 pm
localhost_5100_user_register - Google Chrome 07_03_2022 7_00_06 pm

### References
[Resolving the ActionController::Invalid AuthenticityToken Error in a Rails/React App](https://dev.to/lizlaffitte/resolving-the-actioncontroller-invalidauthenticitytoken-error-in-a-rails-react-app-17m6)
[React-Rails authentication API with Devise and Devise-jwt](https://devsday.ru/blog/details/52846)

### CODE REVIEW

File: app/controllers/admins/registrations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  def register_success
+    render json: { message: 'Signed up sucessfully.' }
+  end
+
+  def register_failed
+    render json: { message: "Something went wrong." }
+  end
+end
```
Comment (analyn-cajocson) [inline app/controllers/admins/registrations_controller.rb:20]: add new line at the end of file

File: app/controllers/admins/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  end
+
+  def log_out_success
+    render json: { message: "You are logged out." }, status: :ok
+  end
+
+  def log_out_failure
+    render json: { message: "Hmm nothing happened."}, status: :unauthorized
+  end
+end
```
Comment (analyn-cajocson) [inline app/controllers/admins/sessions_controller.rb:29]: add new line at the end of file

File: app/controllers/members_controller.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class MembersController < ApplicationController
+  before_action :authenticate_user!
+
+  def show
+    render json: { message: "Yeppa you did it" }
+  end
+end
```
Comment (analyn-cajocson) [inline app/controllers/members_controller.rb:7]: add new line at the end of file

File: app/controllers/users/confirmations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+      puts resource.errors
+    end
+
+  end
+
+private
+  def after_confirmation_path_for(resource_name, resource)
+  end
+
+end
```
Comment (analyn-cajocson) [inline app/controllers/users/confirmations_controller.rb:18]: add new line at the end of file

File: app/controllers/users/registrations_controller.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def register_success
+    render json: { message: 'Signed up sucessfully.' }
+  end
+
+  def register_failed(errors)
+    render json: { message: "Something went wrong.", errors: errors }
+  end
+end
```
Comment (analyn-cajocson) [inline app/controllers/users/registrations_controller.rb:21]: add new line at the end of file

File: app/controllers/users/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  end
+
+  def log_out_success
+    render json: { message: "You are logged out." }, status: :ok
+  end
+
+  def log_out_failure
+    render json: { message: "Hmm nothing happened."}, status: :unauthorized
+  end
+end
```
Comment (analyn-cajocson) [inline app/controllers/users/sessions_controller.rb:29]: add new line at the end of file

File: app/javascript/packs/components/AuthExtension.jsx [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    <Paper sx={{ display: 'flex', alignItems: 'center', justifyContent: 'center', flexDirection: 'row', width: '50%', height: '16rem', textAlign: 'center', bgcolor: 'success'}}>
+      <Box>
+        <Typography variant="h5" component="h1">
+          { isRegister ? 'Have an account?' : 'No account yet?' }
+        </Typography>
+        <Button component={Link} variant="contained" to={ isRegister ? '/user/login' : '/user/register' } sx={{ borderRadius: 24, px: '1rem', my: '.6rem' }} color="success">{ isRegister ? 'Login' : 'Register' }</Button>
+      </Box>
+    </Paper>
+  )
+}
```
Comment (analyn-cajocson) [inline app/javascript/packs/components/AuthExtension.jsx:23]: add new line at the end of file

File: app/javascript/packs/components/Login.jsx [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        <Button variant="contained" type="submit" color="primary" sx={{ width: "100%", my: ".4rem" }}>Login</Button>
+
+
+
+      </Paper>
+
+      <AuthExtension />
+    </>
+  )
+}
```
Comment (analyn-cajocson) [inline app/javascript/packs/components/Login.jsx:60]: add new line at the end of file

File: app/javascript/packs/components/Register.jsx [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        </Box>
+
+        <Button variant="contained" type="submit" color="primary" sx={{ width: "100%", my: ".4rem" }}>Register</Button>
+
+      </Paper>
+
+      <AuthExtension isRegister={true} />
+    </>
+  )
+}
```
Comment (analyn-cajocson) [inline app/javascript/packs/components/Register.jsx:63]: add new line at the end of file

File: app/javascript/packs/components/SideBar.jsx [hunk 10]
```diff
@@ -0,0 +1,5 @@
+import React from 'react'
+
+export default function SideBar() {
+  
+}
```
Comment (analyn-cajocson) [inline app/javascript/packs/components/SideBar.jsx:5]: add new line at the end of file

File: app/javascript/packs/config/theme.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    },
+    secondary: {
+      main: '#f50057',
+    },
+    background: {
+      default: '#141519',
+      paper: '#1e1f24',
+    },
+  },
+});
```
Comment (analyn-cajocson) [inline app/javascript/packs/config/theme.js:17]: add new line at the end of file

File: app/javascript/packs/index.css [hunk 12]
```diff
@@ -0,0 +1,3 @@
+body {
+  margin: 0;
+}
```
Comment (analyn-cajocson) [inline app/javascript/packs/index.css:3]: add new line at the end of file

File: app/javascript/packs/pages/App.js [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+export default function App() {
+  return (
+    <ThemeProvider theme={themeOptions}>
+      <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', bgcolor: 'background.default', height: '100vh' }}>
+        <Outlet />
+
+      </Box>
+    </ThemeProvider>
+  )
+}
```
Comment (analyn-cajocson) [inline app/javascript/packs/pages/App.js:18]: add new line at the end of file

File: app/javascript/packs/pages/ConfirmRequired.js [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+  return (
+    <Paper sx={{ textAlign: 'center', width: '40%', p: '2rem' }}>
+      <Typography variant="h4" component="h1">
+        Confirmation Required
+      </Typography>
+      <p>One more step. Check your email to confirm your account.</p>
+      <Button component={Link} variant="contained" to="/user/login">Log In</Button>
+    </Paper>
+  )
+}
```
Comment (analyn-cajocson) [inline app/javascript/packs/pages/ConfirmRequired.js:21]: add new line at the end of file

File: app/javascript/packs/pages/Dashboard.js [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+
+export default function Dashboard() {
+  return (
+    <div>
+      <h1>Dashboard</h1>
+
+
+    </div>
+  )
+}
```
Comment (analyn-cajocson) [inline app/javascript/packs/pages/Dashboard.js:11]: add new line at the end of file

File: app/javascript/packs/pages/User.js [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+  Box
+} from '@mui/material'
+
+export default function User() {
+  return (
+    <Box elevation={6} sx={{ height: 'auto', width: '50%', display: 'flex', alignItems: 'center' }}>
+      <Outlet />
+    </Box>
+  )
+}
```
Comment (analyn-cajocson) [inline app/javascript/packs/pages/User.js:14]: add new line at the end of file

File: app/models/jwt_denylist.rb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+class JwtDenylist < ApplicationRecord
+  include Devise::JWT::RevocationStrategies::Denylist
+
+  self.table_name = 'jwt_denylist'
+end
```
Comment (analyn-cajocson) [inline app/models/jwt_denylist.rb:5]: add new line at the end of file

File: config/database.yml [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+  <<: *default
+  database: rails_project_development
+
+test:
+  <<: *default
+  database: rails_project_test
+
+production:
+  <<: *default
+  detabase: rails_project_production
```
Comment (analyn-cajocson) [inline config/database.yml:19]: you don't have to include this file on your PR. we already have the `database.yml.sample`

File: db/migrate/20220225034320_create_jwt_denylist.rb [hunk 19]
```diff
@@ -0,0 +1,9 @@
+class CreateJwtDenylist < ActiveRecord::Migration[6.0]
+  def change
+    create_table :jwt_denylist do |t|
+      t.string :jti, null: false
+      t.datetime :exp, null: false
+    end
+    add_index :jwt_denylist, :jti
+  end
+end
```
Comment (analyn-cajocson) [inline db/migrate/20220225034320_create_jwt_denylist.rb:9]: add new line at the end of file

File: spec/requests/admin_spec.rb [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+      post '/admins/sign_in', params: params, as: :json
+      
+      expect(response).to have_http_status(:ok)
+      expect(response.headers['Authorization']).to_not eq nil
+    end
+
+  end
+
+
+end
```
Comment (analyn-cajocson) [inline spec/requests/admin_spec.rb:58]: add new line at the end of file

File: spec/requests/user_spec.rb [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+      post '/users/sign_in', params: params, as: :json
+      
+      expect(response).to have_http_status(:ok)
+      expect(response.headers['Authorization']).to_not eq nil
+    end
+
+  end
+
+
+end
```
Comment (analyn-cajocson) [inline spec/requests/user_spec.rb:58]: add new line at the end of file

**Review by analyn-cajocson — CHANGES_REQUESTED** [review]

File: app/controllers/admins/registrations_controller.rb (hunk 1 above)
Comment (nein-tanuki) [reply app/controllers/admins/registrations_controller.rb]: Adds new line at the end of the file

File: app/controllers/admins/sessions_controller.rb (hunk 2 above)
Comment (nein-tanuki) [reply app/controllers/admins/sessions_controller.rb]: added new line

File: app/controllers/members_controller.rb (hunk 3 above)
Comment (nein-tanuki) [reply app/controllers/members_controller.rb]: added new line

File: app/controllers/users/confirmations_controller.rb (hunk 4 above)
Comment (nein-tanuki) [reply app/controllers/users/confirmations_controller.rb]: added new line

File: app/controllers/users/registrations_controller.rb (hunk 5 above)
Comment (nein-tanuki) [reply app/controllers/users/registrations_controller.rb]: added new line

File: app/models/jwt_denylist.rb (hunk 17 above)
Comment (nein-tanuki) [reply app/models/jwt_denylist.rb]: added new line

File: db/migrate/20220225034320_create_jwt_denylist.rb (hunk 19 above)
Comment (nein-tanuki) [reply db/migrate/20220225034320_create_jwt_denylist.rb]: added new line

File: spec/requests/admin_spec.rb (hunk 20 above)
Comment (nein-tanuki) [reply spec/requests/admin_spec.rb]: added new line

File: spec/requests/user_spec.rb (hunk 21 above)
Comment (nein-tanuki) [reply spec/requests/user_spec.rb]: added new line

File: app/javascript/packs/components/Register.jsx (hunk 9 above)
Comment (nein-tanuki) [reply app/javascript/packs/components/Register.jsx]: added new line

File: app/javascript/packs/components/SideBar.jsx (hunk 10 above)
Comment (nein-tanuki) [reply app/javascript/packs/components/SideBar.jsx]: added new line

File: app/javascript/packs/config/theme.js (hunk 11 above)
Comment (nein-tanuki) [reply app/javascript/packs/config/theme.js]: added new line

File: app/javascript/packs/index.css (hunk 12 above)
Comment (nein-tanuki) [reply app/javascript/packs/index.css]: added new line

File: app/javascript/packs/pages/App.js (hunk 13 above)
Comment (nein-tanuki) [reply app/javascript/packs/pages/App.js]: added new line

File: app/javascript/packs/pages/ConfirmRequired.js (hunk 14 above)
Comment (nein-tanuki) [reply app/javascript/packs/pages/ConfirmRequired.js]: added new line

File: app/javascript/packs/pages/Dashboard.js (hunk 15 above)
Comment (nein-tanuki) [reply app/javascript/packs/pages/Dashboard.js]: added new line

File: app/javascript/packs/pages/User.js (hunk 16 above)
Comment (nein-tanuki) [reply app/javascript/packs/pages/User.js]: added new line

File: config/database.yml (hunk 18 above)
Comment (nein-tanuki) [reply config/database.yml]: renamed database.yml to database.yml.sample
added config/database.yml to .gitignore

**Review by analyn-cajocson — APPROVED** [review]

**Review by analyn-cajocson — CHANGES_REQUESTED** [review]
Fix error first

**Review by analyn-cajocson — APPROVED** [review]

**Review by justinbanaga — APPROVED** [review]

**Pull request merged**